### PR TITLE
Test suite: explicitly ignore irrelevant warnings

### LINKED ISSUE #2161: Regression: Dataset.update(Dataset)
Dataset().update(Dataset())

 

FutureWarning: iteration over an xarray.Dataset will change in xarray v0.11 to only include data variables, not coordinates. Iterate over the Dataset.variables property instead to preserve existing behavior in a forwards compatible manner.

This is a regression in xarray 0.10.4.
@shoyer this isn't serious enough to warrant an immediate release on its own, but we're already doing one so we might as well include it.

### PR DESCRIPTION
**Pull request by shoyer**
Includes silencing two warnings that show up in xarray's public API:

1. `Dataset.update(Dataset)` #2161
2. `.equals()` with scalar timedelta64/datetime64 arrays containing NaT:

```
In [1]: import numpy as np

In [2]: import xarray as xr

In [3]: array = xr.DataArray([np.datetime64('NaT')])

In [4]: array[0].equals(array[0])
/Users/shoyer/dev/xarray/xarray/core/duck_array_ops.py:148: FutureWarning: In the future, 'NAT == x' and 'x == NAT' will always be False.
  flag_array = (arr1 == arr2)
Out[16]: True
```

Fixes #2161

 - [x] Tests added (for all bug fixes or enhancements)
 - [x] Tests passed (for all non-documentation changes)
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API (remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later)

### CODE REVIEW

**Comment by shoyer** [comment]
@fujiisoup any thoughts?

Do you agree that we should have different behavior for overriding `coords` with `__setitem__` vs. `update()`?

**Comment by fujiisoup** [comment]
I didn't notice that iterating over dataset gives a FutureWarning.

> Do you agree that we should have different behavior for overriding coords with __setitem__ vs. update()?

I actually do not yet understand well the difference between `ds.update(other)` and `ds.merge(other)`.

The docs says
> If any dimensions would have inconsistent sizes in the updated dataset.

`update` raises a ValueError, but it does not.
Maybe it should work identically with `xr.merge(other)`?

**Comment by shoyer** [comment]
-[ ] make a decision on coordinate conflicts with `update()`

**Comment by shoyer** [comment]
OK, updated to only silence the warning in `Dataset.update(Dataset)` (by refactoring the logic)

**Comment by fujiisoup** [comment]
Do we update coordinates if other is Dataset?

**Comment by shoyer** [comment]
Yes, if other is a Dataset coordinates get overwritten. We probably already
had test coverage for this, but I also added some explicit tests.

**Comment by shoyer** [comment]
I plan to merge this shortly (with a day or two) unless anyone else has comments

File: xarray/core/merge.py [hunk 1]
```diff
@@ -547,21 +547,24 @@ def dataset_merge_method(dataset, other, overwrite_vars, compat, join):
...
-    other = other.copy()
-    for k, obj in other.items():
-        if isinstance(obj, (Dataset, DataArray)):
-            # drop duplicated coordinates
-            coord_names = [c for c in obj.coords
-                           if c not in obj.dims and c in dataset.coords]
-            if coord_names:
-                other[k] = obj.drop(coord_names)
+    if not isinstance(other, Dataset):
+        other = OrderedDict(other)
```
Comment (fujiisoup) [inline xarray/core/merge.py]: I am slightly wondering this behavior may change after we finally support #884, i.e. we will not include coordinates in `.key()`.

**Comment by shoyer** [comment]
This line only applies to non-Dataset objects (see the isintance check on
the line above), so I think we are OK here.

**Comment by fujiisoup** [comment]
I just missed this line. 
Thanks. 
LGTM.

**Pull request merged**